Cython/Compiler/ParseTreeTransforms.py

```diff
@@ -3114,17 +3127,42 @@ def _inject_eval(self, node, func_name):
                     node.pos, self.current_scope_node(), lenv))
         return node
 
+    def _check_inside_class(self, def_node):
+        return isinstance(def_node, Nodes.DefNode) and any(isinstance(node, Nodes.ClassDefNode) for node, _ in self.env_stack)
+
+    def _get_current_class_and_scope(self):
+        return next(
+            (node, scope)
+            for node, scope
+            in self.env_stack[::-1]
+            if isinstance(node, Nodes.ClassDefNode)
+        )
+    
+    def _inject_class(self, node):
+        # bare __class__ reference inside function
+        def_node = self.current_scope_node()
+
+        if not self._check_inside_class(def_node):
+            return node
+
+        class_node, class_scope = self._get_current_class_and_scope()
+
+        if class_scope.is_py_class_scope:
+            def_node.has_class_reference = True
+            return ExprNodes.NameNode(node.pos, name=class_node.target.name)
+
+        return node
+
     def _inject_super(self, node, func_name):
         lenv = self.current_env()
         entry = lenv.lookup_here(func_name)
         if entry or node.args:
             return node
         # Inject no-args super
         def_node = self.current_scope_node()
-        if (not isinstance(def_node, Nodes.DefNode) or not def_node.args or
-            len(self.env_stack) < 2):
+        if not self._check_inside_class(def_node) or not def_node.args:
```

Comment:
I think it's also worth testing if a method `meth(*args)` satisfies this condition. IIRC, `*args` and `**kwargs` are not part of the `.args` list of the def-node.

Reply:
I might have gone a bit overboard with this, but there's new tests in `py3k_super.pyx` for `@staticmethods` and `@classmethods`.